### PR TITLE
Managed Groups service

### PR DESCRIPTION
**Pull request by jefferai**
This implements the service handler for managed groups, and adds it to the perms table on the website.

CLI and SDK will come in a follow-on PR.

### CODE REVIEW

**Comment by vancluever** [comment]
Looks like we have a failing test:

```
managed_groups chrism$ go test ./
--- FAIL: TestGet (5.55s)
    --- FAIL: TestGet/Get_an_oidc_managed_group (0.01s)
        managed_group_service_test.go:163:
            	Error Trace:	managed_group_service_test.go:163
            	Error:      	Should be empty, but was   (*services.GetManagedGroupResponse)(Inverse(protocmp.Transform, protocmp.Message{
            	            	  	"@type": s"controller.api.services.v1.GetManagedGroupResponse",
            	            	  	"item": protocmp.Message{
            	            	  		... // 5 identical entries
            	            	  		"id":           string("mgoidc_ekZGlew0Jm"),
            	            	  		"scope":        protocmp.Message{"@type": s"controller.api.resources.scopes.v1.ScopeInfo", "id": string("o_J57voY48pe"), "parent_scope_id": string("global"), "type": string("org")},
            	            	- 		"type":         string("oidc"),
            	            	+ 		"type":         string("password"),
            	            	  		"updated_time": protocmp.Message{"@type": s"google.protobuf.Timestamp", "nanos": int32(222428000), "seconds": int64(1622585788)},
            	            	  		"version":      uint32(1),
            	            	  	},
            	            	  }))
            	Test:       	TestGet/Get_an_oidc_managed_group
            	Messages:   	GetManagedGroup("id:\"mgoidc_ekZGlew0Jm\"") got response "item:{id:\"mgoidc_ekZGlew0Jm\" scope:{id:\"o_J57voY48pe\" type:\"org\" parent_scope_id:\"global\"} created_time:{seconds:1622585788 nanos:222428000} updated_time:{seconds:1622585788 nanos:222428000} version:1 type:\"oidc\" auth_method_id:\"amoidc_WCmNLOC2rd\" attributes:{fields:{key:\"filter\" value:{string_value:\"\\\"/foo\\\" == \\\"bar\\\"\"}}} authorized_actions:\"no-op\" authorized_actions:\"read\" authorized_actions:\"update\" authorized_actions:\"delete\"}", wanted "item:{id:\"mgoidc_ekZGlew0Jm\" scope:{id:\"o_J57voY48pe\" type:\"org\" parent_scope_id:\"global\"} created_time:{seconds:1622585788 nanos:222428000} updated_time:{seconds:1622585788 nanos:222428000} version:1 type:\"password\" auth_method_id:\"amoidc_WCmNLOC2rd\" attributes:{fields:{key:\"filter\" value:{string_value:\"\\\"/foo\\\" == \\\"bar\\\"\"}}} authorized_actions:\"no-op\" authorized_actions:\"read\" authorized_actions:\"update\" authorized_actions:\"delete\"}"
FAIL
FAIL	github.com/hashicorp/boundary/internal/servers/controller/handlers/managed_groups	40.687s
FAIL
```

File: internal/servers/controller/handlers/managed_groups/managed_group_service_test.go [hunk 1]
```diff
@@ -0,0 +1,987 @@
...
+	omg := oidc.TestManagedGroup(t, conn, oidcAm, testFakeManagedGroupFilter)
+
+	oidcWireManagedGroup := pb.ManagedGroup{
+		Id:                omg.GetPublicId(),
+		AuthMethodId:      omg.GetAuthMethodId(),
+		CreatedTime:       omg.GetCreateTime().GetTimestamp(),
+		UpdatedTime:       omg.GetUpdateTime().GetTimestamp(),
+		Scope:             &scopepb.ScopeInfo{Id: org.GetPublicId(), Type: scope.Org.String(), ParentScopeId: scope.Global.String()},
+		Version:           1,
+		Type:              "password",
```
Comment (vancluever) [inline internal/servers/controller/handlers/managed_groups/managed_group_service_test.go:118]: Think the failing test might just be a broken fixture?

```suggestion
		Type:              "oidc",
```

File: internal/servers/controller/handlers/managed_groups/managed_group_service_test.go [hunk 2]
```diff
@@ -0,0 +1,987 @@
...
+	s, err := managed_groups.NewService(oidcRepoFn)
+	require.NoError(err, "Error when getting new user service")
+	req := &pbs.DeleteManagedGroupRequest{
+		Id: oidcMg.GetPublicId(),
+	}
+	_, gErr := s.DeleteManagedGroup(auth.DisabledAuthTestContext(iamRepoFn, o.GetPublicId()), req)
+	assert.NoError(gErr, "First attempt")
+	_, gErr = s.DeleteManagedGroup(auth.DisabledAuthTestContext(iamRepoFn, o.GetPublicId()), req)
+	assert.Error(gErr, "Second attempt")
+	assert.True(errors.Is(gErr, handlers.ApiErrorWithCode(codes.NotFound)), "Expected permission denied for the second delete.")
```
Comment (vancluever) [inline internal/servers/controller/handlers/managed_groups/managed_group_service_test.go:422]:
```suggestion
	assert.True(errors.Is(gErr, handlers.ApiErrorWithCode(codes.NotFound)), "Expected not found for the second delete.")
```

**Review by vancluever — COMMENTED** [review]
Generally looking good! I think the failing test is just due to an error in the fixture (suggestion included).

One thing that I did notice on a coverage report is that there's a curious amount of missed code in some of the repo functions where I would have expected them to be covered via some of the tests, mostly associated with `NotFound`-type errors... maybe a quick review is in order?

File: internal/servers/controller/handlers/managed_groups/managed_group_service.go [hunk 3]
```diff
@@ -0,0 +1,661 @@
...
+}
+
+// Service handles request as described by the pbs.ManagedGroupServiceServer interface.
+type Service struct {
+	pbs.UnimplementedManagedGroupServiceServer
+
+	oidcRepoFn common.OidcAuthRepoFactory
+}
+
+// NewService returns a user service which handles user related requests to boundary.
```
[the remark this answers is not in the document]
Comment (jefferai) [reply internal/servers/controller/handlers/managed_groups/managed_group_service.go:75]: Hah -- I made sure to find all the places where accounts were mentioned -- but this was a place that was already wrong when copied and pasted into the accounts service! I'll fix it there too.

File: internal/proto/local/controller/api/services/v1/managed_group_service.proto [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+  // UpdateManagedGroup updates an existing ManagedGroup in boundary. The
+  // provided ManagedGroup must not have any read only fields set. The update
+  // mask must be included in the request and contain at least 1 mutable field.
+  // To unset a field's value, include the field in the update mask and don't
+  // set it in the provided ManagedGroup. An error is returned if the
+  // ManagedGroup id is missing or references a non-existing resource. An error
+  // is also returned if the request attempts to update the name to one that is
+  // already in use in the containing Auth Method.
+  rpc UpdateManagedGroup(UpdateManagedGroupRequest) returns (UpdateManagedGroupResponse) {
```
[the remark this answers is not in the document]
Comment (jefferai) [reply internal/proto/local/controller/api/services/v1/managed_group_service.proto:66]: That's part of our general API design for all resources...we don't make that explicit in the comments for other services so I don't think we need to here.

File: internal/servers/controller/handlers/managed_groups/validate_test.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+			name: "unrecognized authmethod prefix",
+			item: &pb.ManagedGroup{
+				AuthMethodId: "anything_1234567890",
+			},
+			errContains: fieldError(authMethodIdField, "Unknown auth method type from ID."),
+		},
+		{
+			name: "mismatched oidc authmethod pw type",
+			item: &pb.ManagedGroup{
+				Type:         "ldap",
```
Comment (jimlambrt) [inline internal/servers/controller/handlers/managed_groups/validate_test.go:38]: Oh how I miss the good old days of openLDAP.  Life and IAM was so much simpiler back then.

File: internal/servers/controller/handlers/managed_groups/validate_test.go [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+	"github.com/stretchr/testify/require"
+	"google.golang.org/protobuf/types/known/fieldmaskpb"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+func fieldError(field, details string) string {
+	return fmt.Sprintf(`{name: %q, desc: %q}`, field, details)
+}
+
+func TestValidateCreateRequest(t *testing.T) {
```
Comment (jimlambrt) [inline internal/servers/controller/handlers/managed_groups/validate_test.go:22]: any reason why we shouldn't run the tests in this file in parallel?

File: internal/servers/controller/handlers/managed_groups/managed_group_service_test.go [hunk 7]
```diff
@@ -0,0 +1,987 @@
...
+	s, err := managed_groups.NewService(oidcRepoFn)
+	require.NoError(t, err, "Couldn't create new managed groups service.")
+
+	org, _ := iam.TestScopes(t, iam.TestRepo(t, conn, wrap))
+
+	databaseWrapper, err := kmsCache.GetWrapper(ctx, org.PublicId, kms.KeyPurposeDatabase)
+	require.NoError(t, err)
+	oidcAm := oidc.TestAuthMethod(
+		t, conn, databaseWrapper, org.PublicId, oidc.ActivePrivateState,
+		"alice-rp", "fido",
```
Comment (jimlambrt) [inline internal/servers/controller/handlers/managed_groups/managed_group_service_test.go:104]: seeing Alice's dog's name being cut/pasted as a test client_secret makes me happy!

File: internal/servers/controller/handlers/managed_groups/managed_group_service.go [hunk 8]
```diff
@@ -0,0 +1,661 @@
...
+	attributesField   = "attributes"
+	filterField       = "filter"
+	idField           = "id"
+
+	// oidc field names
+	attrFilterField = "attributes.filter"
+)
+
+var (
+	pwMaskManager   handlers.MaskManager
```
Comment (jimlambrt) [inline internal/servers/controller/handlers/managed_groups/managed_group_service.go:39]: I'm not sure this is need and/or used.

File: internal/servers/controller/handlers/managed_groups/managed_group_service.go [hunk 9]
```diff
@@ -0,0 +1,661 @@
...
+	if rowsUpdated == 0 {
+		return nil, handlers.NotFoundErrorf("Managed Group %q doesn't exist or incorrect version provided.", id)
+	}
+	return out, nil
+}
+
+func (s Service) updateInRepo(ctx context.Context, scopeId, authMethodId string, req *pbs.UpdateManagedGroupRequest) (auth.ManagedGroup, error) {
+	const op = "managed_groups.(Service).updateInRepo"
+	var out auth.ManagedGroup
+	switch auth.SubtypeFromId(req.GetId()) {
```
Comment (jimlambrt) [inline internal/servers/controller/handlers/managed_groups/managed_group_service.go:402]: ty for using the existing subtype pattern here, even though there's currently only one type of managed group.

File: internal/proto/local/controller/api/services/v1/managed_group_service.proto [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+  rpc ListManagedGroups(ListManagedGroupsRequest) returns (ListManagedGroupsResponse) {
+    option (google.api.http) = {
+      get: "/v1/managed-groups"
+     };
+    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
+      summary: "Lists all ManagedGroups in a specific Auth Method."
+    };
+  }
+
+  // CreateManagedGroup creates and stores an ManagedGroup in boundary. The provided
```
Comment (jimlambrt) [inline internal/proto/local/controller/api/services/v1/managed_group_service.proto:41]: Nit: capitalize "Boundary" and I'm not sure the phrase "in Boundary" is needed anywho.

File: internal/proto/local/controller/api/services/v1/managed_group_service.proto [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+      post: "/v1/managed-groups"
+      body: "item"
+      response_body: "item"
+    };
+    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
+      summary: "Creates a single ManagedGroup in the provided Auth Method."
+    };
+  }
+
+  // UpdateManagedGroup updates an existing ManagedGroup in boundary. The
```
Comment (jimlambrt) [inline internal/proto/local/controller/api/services/v1/managed_group_service.proto:58]: same nit about "in boundary"

File: internal/proto/local/controller/api/services/v1/managed_group_service.proto [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+      patch: "/v1/managed-groups/{id}"
+      body: "item"
+      response_body: "item"
+    };
+    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
+      summary: "Updates a ManagedGroup."
+    };
+  }
+
+  // DeleteManagedGroup removes an ManagedGroup from Boundary. If the provided ManagedGroup Id
```
Comment (jimlambrt) [inline internal/proto/local/controller/api/services/v1/managed_group_service.proto:77]: At least "Boundary" is capitalized :)

**Review by jimlambrt — CHANGES_REQUESTED** [review]
Amazing work!  Just one minor thing that's not used and a few doc nits.

File: internal/servers/controller/handlers/managed_groups/validate_test.go (hunk 6 above)
Comment (jefferai) [reply internal/servers/controller/handlers/managed_groups/validate_test.go:22]: Probably not. I can update the account tests to do so as well.

File: internal/servers/controller/handlers/managed_groups/managed_group_service.go (hunk 9 above)
Comment (jefferai) [reply internal/servers/controller/handlers/managed_groups/managed_group_service.go:402]: 👍

File: internal/servers/controller/handlers/managed_groups/managed_group_service.go (hunk 8 above)
Comment (jefferai) [reply internal/servers/controller/handlers/managed_groups/managed_group_service.go:39]: Fixed

File: internal/proto/local/controller/api/services/v1/managed_group_service.proto (hunk 10 above)
Comment (jefferai) [reply internal/proto/local/controller/api/services/v1/managed_group_service.proto:41]: I removed it

File: internal/proto/local/controller/api/services/v1/managed_group_service.proto (hunk 11 above)
Comment (jefferai) [reply internal/proto/local/controller/api/services/v1/managed_group_service.proto:58]: Fixed

**Review by jimlambrt — APPROVED** [review]

**Review by vancluever — APPROVED** [review]
:shipit:

**Comment by jefferai** [comment]
Discussed the coverage stuff with @vancluever offline, we're good for now.

**Pull request merged**